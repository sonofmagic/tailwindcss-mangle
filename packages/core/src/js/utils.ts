import type babel from '@babel/core'
// const t = babel.types
export function getStringLiteralCalleeName(path: babel.NodePath<babel.types.StringLiteral>) {
  if (path.parentPath.isCallExpression()) {
    const callee = path.parentPath.get('callee')
    if (callee.isIdentifier()) {
      return callee.node.name
    }
  }
}

export function getTemplateElementCalleeName(path: babel.NodePath<babel.types.TemplateElement>) {
  if (path.parentPath.isTemplateLiteral()) {
    const pp = path.parentPath
    if (pp.parentPath.isCallExpression()) {
      const callee = pp.parentPath.get('callee')
      if (callee.isIdentifier()) {
        return callee.node.name
      }
    }
  }
}
