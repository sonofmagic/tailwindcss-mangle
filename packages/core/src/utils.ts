import micromatch from 'micromatch'

const { isMatch } = micromatch

export function escapeStringRegexp(str: string) {
  if (typeof str !== 'string') {
    throw new TypeError('Expected a string')
  }
  return str.replaceAll(/[$()*+.?[\\\]^{|}]/g, '\\$&').replaceAll('-', '\\x2d')
}

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (pattern === undefined) {
    return function () {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export { defaultMangleClassFilter, isMap, isRegexp } from '@tailwindcss-mangle/shared'
