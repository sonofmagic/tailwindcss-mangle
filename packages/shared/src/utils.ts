export const preserveClassNames = [
  // https://tailwindcss.com/docs/transition-timing-function start
  // https://github.com/sonofmagic/tailwindcss-mangle/issues/21
  'ease-out',
  'ease-linear',
  'ease-in',
  'ease-in-out'
  // https://tailwindcss.com/docs/transition-timing-function end
]

// eslint-disable-next-line unicorn/no-array-reduce
const preserveClassNamesMap = preserveClassNames.reduce<Record<(typeof preserveClassNames)[number], true>>((acc, cur) => {
  acc[cur] = true
  return acc
}, {})

export const defaultMangleClassFilter = (className: string) => {
  if (preserveClassNamesMap[className]) {
    return false
  }
  // ignore className like 'filter','container'
  // it may be dangerous to mangle/rename all StringLiteral , so use /-/ test for only those with /-/ like:
  // bg-[#123456] w-1 etc...
  return /[:-]/.test(className)
}

export function groupBy<T>(arr: T[], cb: (arg: T) => string): Record<string, T[]> {
  if (!Array.isArray(arr)) {
    throw new TypeError('expected an array for first argument')
  }

  if (typeof cb !== 'function') {
    throw new TypeError('expected a function for second argument')
  }

  const result: Record<string, T[]> = {}
  for (const item of arr) {
    const bucketCategory = cb(item)
    const bucket = result[bucketCategory]

    if (Array.isArray(bucket)) {
      result[bucketCategory].push(item)
    } else {
      result[bucketCategory] = [item]
    }
  }

  return result
}

export const acceptChars = [...'abcdefghijklmnopqrstuvwxyz']

export function stripEscapeSequence(words: string) {
  return words.replaceAll('\\', '')
}

export function isRegexp(value: unknown) {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isMap(value: unknown) {
  return Object.prototype.toString.call(value) === '[object Map]'
}

export function regExpTest(arr: (string | RegExp)[] = [], str: string) {
  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (typeof item === 'string') {
        if (item === str) {
          return true
        }
      } else if (isRegexp(item)) {
        item.lastIndex = 0
        if (item.test(str)) {
          return true
        }
      }
    }
    return false
  }
  throw new TypeError("paramater 'arr' should be a Array of Regexp | String !")
}
