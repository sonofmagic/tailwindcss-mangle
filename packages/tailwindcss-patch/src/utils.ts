import fss from 'node:fs'
import fs from 'node:fs/promises'
import type { SyncOpts } from 'resolve'
import pkg from 'resolve'

const { sync } = pkg

export function ensureFileContent(filepaths: string | string[]) {
  if (typeof filepaths === 'string') {
    filepaths = [filepaths]
  }
  let content
  for (const filepath of filepaths) {
    if (fss.existsSync(filepath)) {
      content = fss.readFileSync(filepath, {
        encoding: 'utf8',
      })
      break
    }
  }
  return content
}

export function requireResolve(id: string, opts?: SyncOpts) {
  return sync(id, opts)
}

export async function ensureDir(p: string) {
  try {
    await fs.access(p)
  }
  catch {
    await fs.mkdir(p, {
      recursive: true,
    })
  }
}
