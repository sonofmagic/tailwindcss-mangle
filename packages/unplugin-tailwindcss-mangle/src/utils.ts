import fs from 'node:fs'
import path from 'node:path'
import micromatch from 'micromatch'

import { groupBy } from '@tailwindcss-mangle/shared'
import { pluginName } from './constants'
const { isMatch } = micromatch

export function getGroupedEntries<T>(
  entries: [string, T][],
  options = {
    cssMatcher(file: string) {
      return /\.css$/.test(file)
    },
    htmlMatcher(file: string) {
      return /\.html?$/.test(file)
    },
    jsMatcher(file: string) {
      return /\.[cm]?js$/.test(file)
    }
  }
) {
  const { cssMatcher, htmlMatcher, jsMatcher } = options
  const groupedEntries = groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return 'css'
    } else if (htmlMatcher(file)) {
      return 'html'
    } else if (jsMatcher(file)) {
      return 'js'
    } else {
      return 'other'
    }
  })
  if (!groupedEntries.css) {
    groupedEntries.css = []
  }
  if (!groupedEntries.html) {
    groupedEntries.html = []
  }
  if (!groupedEntries.js) {
    groupedEntries.js = []
  }
  if (!groupedEntries.other) {
    groupedEntries.other = []
  }
  return groupedEntries as Record<'css' | 'html' | 'js' | 'other', [string, T][]>
}

export function createGlobMatcher(pattern: string | string[] | undefined, fallbackValue: boolean = false) {
  if (pattern === undefined) {
    return function () {
      return fallbackValue
    }
  }
  return function (file: string) {
    return isMatch(file, pattern)
  }
}

export function getCacheDir(basedir = process.cwd()) {
  return path.resolve(basedir, 'node_modules/.cache', pluginName)
}

export function mkCacheDirectory(cwd = process.cwd()) {
  const cacheDirectory = getCacheDir(cwd)

  const exists = fs.existsSync(cacheDirectory)
  if (!exists) {
    fs.mkdirSync(cacheDirectory, {
      recursive: true
    })
  }
  return cacheDirectory
}

export function cacheDump(filename: string, data: any[] | Set<any>, basedir?: string) {
  try {
    const dir = mkCacheDirectory(basedir)
    fs.writeFileSync(path.resolve(dir, filename), JSON.stringify([...data], undefined, 2), 'utf8')
  } catch (error) {
    console.log(error)
  }
}

export { defaultMangleClassFilter, isMap, isRegexp } from '@tailwindcss-mangle/shared'
