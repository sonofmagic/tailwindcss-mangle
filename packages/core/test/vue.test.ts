import { parse } from '@vue/compiler-sfc'
import type {
  NodeTransform,
  RootNode,
} from '@vue/compiler-core'
import {
  CREATE_VNODE,
  RESOLVE_DIRECTIVE,
  baseParse,
  createSimpleExpression,
  generate,
  helperNameMap,
  locStub,
  transform,
} from '@vue/compiler-core'
import { getTestCase } from './utils'

const NodeTypes = {
  ROOT: 0,
  ELEMENT: 1,
  TEXT: 2,
  COMMENT: 3,
  SIMPLE_EXPRESSION: 4,
  INTERPOLATION: 5,
  ATTRIBUTE: 6,
  DIRECTIVE: 7,
  COMPOUND_EXPRESSION: 8,
  IF: 9,
  IF_BRANCH: 10,
  FOR: 11,
  TEXT_CALL: 12,
  VNODE_CALL: 13,
  JS_CALL_EXPRESSION: 14,
  JS_OBJECT_EXPRESSION: 15,
  JS_PROPERTY: 16,
  JS_ARRAY_EXPRESSION: 17,
  JS_FUNCTION_EXPRESSION: 18,
  JS_CONDITIONAL_EXPRESSION: 19,
  JS_CACHE_EXPRESSION: 20,
  JS_BLOCK_STATEMENT: 21,
  JS_TEMPLATE_LITERAL: 22,
  JS_IF_STATEMENT: 23,
  JS_ASSIGNMENT_EXPRESSION: 24,
  JS_SEQUENCE_EXPRESSION: 25,
  JS_RETURN_STATEMENT: 26,
} as const

function createRoot(options: Partial<RootNode> = {}): RootNode {
  return {
    type: 0, // NodeTypes.ROOT,
    children: [],
    helpers: new Set(),
    components: [],
    directives: [],
    imports: [],
    hoists: [],
    cached: 0,
    temps: 0,
    codegenNode: createSimpleExpression(`null`, false),
    loc: locStub,
    ...options,
  }
}
// https://github.com/vuejs/core/blob/main/packages/compiler-sfc/__tests__/templateTransformAssetUrl.spec.ts
describe('vue', () => {
  it('module mode preamble', () => {
    const root = createRoot({
      helpers: new Set([CREATE_VNODE, RESOLVE_DIRECTIVE]),
    })
    const { code } = generate(root, { mode: 'module' })
    expect(code).toMatch(
      `import { ${helperNameMap[CREATE_VNODE]} as _${helperNameMap[CREATE_VNODE]}, ${helperNameMap[RESOLVE_DIRECTIVE]} as _${helperNameMap[RESOLVE_DIRECTIVE]} } from "vue"`,
    )
    expect(code).toMatchSnapshot()
  })
  it.skip('test for vue', () => {
    const testCase = getTestCase('preserve-fn-case1.vue')
    const result = parse(testCase)

    console.log(result.descriptor.template, result.descriptor.script, result.descriptor.scriptSetup)
    const content = result.descriptor.template?.content ?? ''
    const ast = baseParse(content)
    const plugin: NodeTransform = (node, context) => {
      if (node.type === NodeTypes.ELEMENT && node.tag === 'div') {
        // change the node to <p>
        context.replaceNode(
          Object.assign({}, node, {
            tag: 'p',
            children: [
              {
                type: NodeTypes.TEXT,
                content: 'hello',
                isEmpty: false,
              },
            ],
          }),
        )
      }
    }
    transform(ast, {
      nodeTransforms: [plugin],
    })
    const t = generate(ast, { mode: 'module' })
    expect(t.code).toBe(content)
  })
})
