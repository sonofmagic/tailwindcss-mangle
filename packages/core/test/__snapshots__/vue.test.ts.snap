// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vue > module mode preamble 1`] = `
"import { createVNode as _createVNode, resolveDirective as _resolveDirective } from \\"vue\\"

export function render(_ctx, _cache) {
  return null
}"
`;
